### PR TITLE
Add checks for the signature of the `start` lang item

### LINKED ISSUE #105963: Verify signature of start lang item
In https://github.com/bjorn3/rustc_codegen_cranelift/issues/1329 it was found if the signature of the start lang item is wrong, rustc with the LLVM backend will silently compile it despite the ABI mismatch. Cg_clif crashes entirely. To avoid both issues rustc should validate the signature.

### PR DESCRIPTION
**Pull request by asquared31415**
Closes #105963

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/errors.rs [hunk 1]
```diff
@@ -172,3 +172,68 @@ impl AddToDiagnostic for TypeMismatchFruTypo {
...
+    pub found_param_count: usize,
+}
+
+#[derive(Diagnostic)]
+#[diag(hir_typeck_lang_start_incorrect_param)]
+pub struct LangStartIncorrectParam<'tcx> {
+    #[primary_span]
+    param_span: Span,
+
+    #[suggestion(style = "short", code = "{expected_ty}")]
```
Comment (davidtwco) [inline compiler/rustc_hir_typeck/src/errors.rs:192]: You can include the applicability in this attribute, e.g. `#[suggestion(applicability = "machine-applicable", ...)] - might let you remove the constructors.

**Review by davidtwco — APPROVED** [review]
LGTM, r=me after fixing one nit. Apologies for the delay in getting to this review.

**Comment by compiler-errors** [comment]
@bors r-

`src/tools/clippy/tests/ui/def_id_nocore.rs` needs fixing.

**Comment by dtolnay** [comment]
This needs to be rebased past #106458, since it is showing not mergeable in bors.
https://bors.rust-lang.org/queue/rust
@bors r-
@rustbot author

**Comment by dtolnay** [comment]
Here is a rebase on current master: 337a97d3741bb78319cde745119cf96941b78325

**Pull request merged**